src/diagnostics.jl

```diff
@@ -55,25 +52,174 @@ function velocity_div!(grid::RegularCartesianGrid, u, v, w, div)
     end
 end
 
-function run_diagnostic(model::Model, diag::VelocityDivergenceChecker)
-    u, v, w = model.velocities.u.data, model.velocities.v.data, model.velocities.w.data
-    div = model.stepper_tmp.fC1
+@inline _prod(data, ::Nothing, i, j, k) = @inbounds data[i, j, k]
+@inline _prod(data1,    data2, i, j, k) = @inbounds data1[i, j, k] * data2[i, j, k]
+
+# Work-inefficient inclusive scan that reduces along the xy dimensions.
+# Useful for computing mean(; dim=[1, 2]) lol.
+# Modified from: https://github.com/JuliaGPU/CUDAnative.jl/blob/master/examples/scan.jl
+@hascuda function gpu_accumulate_xy!(Rxy::CuDeviceArray{T}, Rx, data1, data2, op::Function) where T
+    lvl, lvls = blockIdx().y,  gridDim().y
+    col, cols = blockIdx().x,  gridDim().x
+    row, rows = threadIdx().x, blockDim().x
+
+    if lvl <= lvls && col <= cols && row <= rows
+        shmem = @cuDynamicSharedMem(T, 2*rows)
+        shmem[row] = _prod(data1, data2, row, col, lvl)
+        sync_threads()
+
+        # parallel reduction
+        pin, pout, offset = 1, 0, 1
+        while offset < rows
+            pout = 1 - pout
+            pin = 1 - pin
+
+            if row > offset
+                shmem[pout * rows + row] = op(shmem[pin * rows + row], shmem[pin * rows + row - offset])
+            else
+                 shmem[pout * rows + row] = shmem[pin * rows + row]
+            end
+
+            sync_threads()
+            offset *= UInt32(2)
+        end
+
+        shmem[pin * rows + row] = shmem[pout * rows + row]
+        sync_threads()
 
-    velocity_div!(model.grid, u, v, w, div)
-    min_div, mean_div, max_div = minimum(div), mean(div), maximum(div)
+        # write results
+        if row == rows
+            Rx[1, col, lvl] = shmem[row]
+        end
+        sync_threads()
 
-    if max(abs(min_div), abs(max_div)) >= diag.warn_threshold
-        t, i = model.clock.time, model.clock.iteration
-        println("time = $t, iteration = $i")
-        println("WARNING: Velocity divergence is high! min=$min_div, mean=$mean_div, max=$max_div")
+        if col == cols
+            sum = 0
+            for j in 1:cols
+                sum += Rx[1, j, lvl]
+            end
+            Rxy[1, 1, lvl] = real(sum)

```

Comment:
Why use `real` here?

Reply:
Because `eltype(solver.storage) == Complex{Float64}`. So it's another hack because we're recycling `solver.storage` =/

Without `real(sum)` I get an obscure GPU error.

Reply:
Ok --- does this mean that `Rx` is some kind of necessary scratch space in this algorithm? And the only scratch space we have to use is `poisson_solver.storage`?